spatial/r3/triangle.go

```diff
@@ -0,0 +1,116 @@
+// Copyright ©2022 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package r3
+
+import "math"
+
+// Triangle represents a triangle in 3D space and
+// is composed by 3 vectors corresponding to the position
+// of each of the vertices. Ordering of these vertices
+// decides the "normal" direction.
+// Inverting ordering of two vertices inverts the resulting direction.
+type Triangle [3]Vec
+
+// Normal returns the vector with direction
+// perpendicular to the Triangle's face. The ordering
+// of the triangle vertices decides the normal's resulting
+// direction. Normal is not guaranteed to return a unit-norm vector.
+func (t Triangle) Normal() Vec {
+	s1, s2, _ := t.sides()
+	return Cross(s1, s2)
+}
+
+// Degenerate returns true if all of triangle's vertices are
+// within tol distance of its longest side.
+func (t Triangle) Degenerate(tol float64) bool {
+	longIdx := t.longIdx()
+	// calculate vertex distance from longest side
+	ln := line{t[longIdx], t[(longIdx+1)%3]}
+	dist := ln.distance(t[(longIdx+2)%3])
+	return dist <= tol
+}
+
+// Area returns the surface area of the triangle.
+func (t Triangle) Area() float64 {
+	// Formula taken from Section 2, page 3.
+	// William M. Kahan (24 March 2000). "Miscalculating Area and Angles of a Needle-like Triangle"
+	// https://people.eecs.berkeley.edu/~wkahan/Triangle.pdf
```

Comment:
Note that this is Heron's formula. I would probably link to the first (as well as this reference) given the ancient history of it. https://en.wikipedia.org/wiki/Heron%27s_formula.

```
// Heron's Formula, see https://en.wikipedia.org/wiki/Heron%27s_formula.
// Also see William M. Kahan (24 March 2000). "Miscalculating Area and Angles of a Needle-like Triangle"
// for more discussion. https://people.eecs.berkeley.edu/~wkahan/Triangle.pdf.
```